Added option hostname to docker service create

vasil-yordanov opened the pull request:
This pull request is related to issue: https://github.com/docker/docker/issues/24877
I have added new field "hostname" in the ContainerSpec, this is needed for hostname option of  docker service create command.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "master" git@github.com:vasil-yordanov/swarmkit.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

vasil-yordanov commented:
Execute on my machine the cd "$WORKDIR" && make ci and it is wokring. How to proceed?

aaronlehmann commented:
@vasil-yordanov: I reran the tests and they passed this time. It looks like an unrelated test failed.

ping @diogomonica to look at the failure: https://circleci.com/gh/docker/swarmkit/3937

```
--- FAIL: TestRenewTLSConfigWithNoNode (2.28s)
	Error Trace:	config_test.go:417
	Error:		TestRenewTLSConfig timed-out
```

aaronlehmann commented:
Opened #1321 to fix the flaky test.

stevvooe commented:
@vasil-yordanov So, it looks like the intent of #24877 is to provide a set of hostnames _injected_ into a container, whereas, judging by this just being a field, this would only set the hostname seen internally by the container the same for each container. Is that the intended behavior?

vasil-yordanov commented:
@stevvooe you are correct this is the behaviour that I have implemented. The rest of the impl is in docker/docker and docker/engine-api repos where I have other pending pull requests.

vasil-yordanov commented:
I have amend the commit so that to sign the first commit

stevvooe commented:
@vasil-yordanov Could you update agent/exec/container to use this field?

I'm not sure that this meets the use case outlined in docker/docker#24877. That issue described something closer to templating of hostnames, albeit, this has a better chance of landing.

vasil-yordanov commented:
@stevvooe done. About the templateing, I have not implemented this, there was a discussion in docker/docker#24877 to use different option like --container-hostname, I think it is still possible to use the same option, hostname, and depending on the value of that option to assume that value as a template.

vasil-yordanov commented:
@stevvooe I have just cloned https://github.com/docker/docker.git . 
Do you have any idea why there is no such folder agent/exec/container in the docker/docker repo into src/github.com/docker/swarmkit/ i can see only agent/exec?

stevvooe commented:
@vasil-yordanov This is in swarmkit. The executor in docker/docker is in docker/daemon/cluster/executor/container. Both need to be changed. We use the former for testing and the latter is the production version.

stevvooe commented on agent/exec/container/container.go:
```diff
@@ -83,9 +83,11 @@ func (c *containerConfig) image() string {
 }
 
 func (c *containerConfig) config() *enginecontainer.Config {
+	var hostname = strings.Replace(c.spec().Hostname, "*", fmt.Sprint(c.task.Slot), 1)
```
We need to be careful about creating a "template" system without thinking about consistent user experience.

See https://github.com/docker/docker/issues/24110#issuecomment-234358152 for wider discussion on this.

I know this is a desired feature but there are a lot implications to whatever choice we make.

vasil-yordanov replied:
Actually I need only fixed hostname on all containers. Would you accept such kind of UX?
I can easily revert my last commit in case fixed host name is acceptable.

Templated hostname will be useful for me if 'source' parameter of 'mount' option of docker service create is also templated.

stevvooe replied:
> Templated hostname will be useful for me if 'source' parameter of 'mount' option of docker service create is also templated.

This is why templating needs more thought. I'd like this to work consistently and safely across the board.

Yes, this PR will be acceptable without the templating.

vasil-yordanov commented:
@stevvooe Ok, I have rebased and remove the commit with the hostname template. It seems that the templateing idea has to grow and has to have broader discussions among the community before to be implemented.

stevvooe commented:
LGTM

@aluzzardi Do we enforce adding swarmkit flags for these kinds of additions?

aluzzardi commented:
@stevvooe CLI flags? Nah. Greatly appreciated though

stevvooe commented:
@vasil-yordanov Please rebase and we can get this merged.

vasil-yordanov commented:
@stevvooe I have rebased. There was a conflict in 
CONFLICT (content): Merge conflict in api/specs.pb.go

I have rerun the 'make generate' and 'make ci' is working locally.
However in the ci/cicleci there is a failure:

```
_👹 please run 'make generate' when making changes to proto files
make: *** [checkprotos] Error 1

cd "$WORKDIR" && make ci returned exit code 2_
```

Do you have any idea why its failing?

aaronlehmann commented:
We upgraded our CI system to use Go 1.7. When it regenerates the protobuf files, the gzip compression appears to be different between Go 1.6 and Go 1.7. So upgrading to Go 1.7 and rerunning `make generate` should solve the problem.

vasil-yordanov commented:
@stevvooe and @aaronlehmann I rebased. Done.

stevvooe commented:
LGTM

This may need some validation.

vasil-yordanov commented:
@stevvooe Should I write the validation code or this will be done by a QA from your team?
Do you have any idea, if this change will be incorporated into the docker/docker repo?
I have created a pull request there, but no body is responding to this pull request.
https://github.com/docker/docker/pull/25437

Today I'm going to rebase my pull request to the docker/docker repo, which is also related to this option but in the docker repo.

stevvooe commented:
> Should I write the validation code or this will be done by a QA from your team?

No, any validation of the hostname parameter accepted by the swarm manager must be a part of this PR. We may require the code in https://github.com/docker/swarmkit/blob/master/manager/controlapi/service.go#L105 to ensure that the value is a valid hostname, but if this is platform specific, this may not be possible.

> Do you have any idea, if this change will be incorporated into the docker/docker repo?
> I have created a pull request there, but no body is responding to this pull request.
docker/docker#25437

Yes, but this PR is required before proceeding. We are not looking at docker/docker#25437 since this needs to be reviewed first. Once this PR Is accepted, you'll update the swarmkit dependencies in hack/vendor.sh to pull in these new changes.

vasil-yordanov commented:
@stevvooe I got it. You were talking about parameter validation. I was thinking about some functional testing (validation test code) of this new feature :) .
Ok, I will add the parameter validation. There should be a RFC for the hostname and fqdn, in general the host name validation should not depend on the platform. Eventhough I had some problem with OSX when I have used a dash in the hostname the Mac was not able to connect to internet :) and it was very difficult to discover root cause of the issue. 

I will check tomorrow how the hostname validation is implemented in docker/docker repo.
There is such option --hostname in 'docker run' command, I will check its validation and will add the validation code to the swarmkit code.

stevvooe commented on manager/controlapi/service.go:
```diff
@@ -102,6 +103,29 @@ func validateUpdate(uc *api.UpdateConfig) error {
+
+	if container.Hostname != "" {
+		// Validate if the given hostname is RFC 1123
+		matched, _ := regexp.MatchString("^(([[:alnum:]]|[[:alnum:]][[:alnum:]\\-]*[[:alnum:]])\\.)*([[:alnum:]]|[[:alnum:]][[:alnum:]\\-]*[[:alnum:]])$", container.Hostname)
```
A few things:

- make a function for this
- pre-compile the regexp (use the global pattern with regexp.MustCompile)

vasil-yordanov commented:
@stevvooe I did the change, with the pre-compiled regesxp, but there is some strange failure of the 'make ci' I'm investigating if it is related to my change.

stevvooe commented on manager/controlapi/service.go:
```diff
@@ -21,6 +22,9 @@ var (
 )
 
+// Regexp pattern for hostname to conform RFC 1123
+var hostnamePattern = regexp.MustCompile("^(([[:alnum:]]|[[:alnum:]][[:alnum:]\\-]*[[:alnum:]])\\.)*([[:alnum:]]|[[:alnum:]][[:alnum:]\\-]*[[:alnum:]])$")
```
@dmcgowan PTAL

stevvooe replied:
@vasil-yordanov Let's see if we can get a unit test with some examples for this.

dmcgowan replied:
Logic of regex looks right, agreed tests should be used to verify. Also non-capturing groups could be used here.

stevvooe commented:
@vasil-yordanov This looks great! I think we need a few test cases around the hostname validation.

Thanks for keeping with this!

vasil-yordanov commented:
@stevvooe I have added some test cases for the hostname validation. There is still one test case failing, but it seems it is not related to my changes.
TestServiceUpdateRejectNetworkChange, do you know if network update is already supported. The failing testcases is around errNetworkUpdateNotSupported, and it seems the UpdateService method does note return error (err is nil).

aaronlehmann commented on api/specs.proto:
```diff
@@ -188,6 +192,7 @@ message ContainerSpec {
 
 	// PullOptions parameterize the behavior of image pulls.
 	PullOptions pull_options = 10;
+
```
nit: extra blank line added

aaronlehmann commented:
@vasil-yordanov: The test failure seems to be caused by this commit in your PR:

```
commit ee0ae334113155c70597ac140b19e9743cc72cc2
Author: Vasil Yordanov <vasil.yordanov@gmail.com>
Date:   Fri Aug 26 17:26:44 2016 +0300

    Added validation for hostname parameter
```

I think there is a rebase issue with this commit. It changes some code in `UpdateService` under the comment `// temporary disable network update`. It looks like it's restoring an older version of this code.

vasil-yordanov commented:
@aaronlehmann I fixed the issue. It seems that it was a rebase issue. Somehow I have gotten an old code in the file service.go that I have changed. Usually I do git diff before commit. Have no idea how I miss this.
 
I have re-written the git history, to have clean and small pull request. Also removed the  empty line in spec.proto, just to have clean code, but this was not the reason of the problem.
However the CI is still failing, in the circleci, but with some very strange error
"--- FAIL: TestRaftSnapshotRestart (10.44s)
	Error Trace:	storage_test.go:183
	Error:		Received unexpected error "context deadline exceeded""

Note that when I execute the 'make ci' on my machine it passes without issues. So I do not know how to solve the circleci issue, when it is working locally for me.

aaronlehmann commented:
> --- FAIL: TestRaftSnapshotRestart (10.44s)
> Error Trace: storage_test.go:183
> Error: Received unexpected error "context deadline exceeded"

This should be fixed by #1469. If you rebase on top of the latest master branch, it should resolve the issue.

vasil-yordanov commented:
@aaronlehmann thanks for the help. Finally I have the circle-ci passed :)

stevvooe commented:
LGTM

Thanks for the great PR! CC me on the docker/docker PR and we'll work from there!

AkihiroSuda commented:
> // All containers for a given service will have the same hostname

Just a question :smiley: , what is your concrete use case for that?

niau commented:
> Just a question 😃 , what is your concrete use case for that?

 Currently the hostname is dynamic and when the container got respawned on another host it will change. I know that it might sound stupid but I know legacy applications that depend a lot on it... So it is actually nice to have.

vasil-yordanov commented:
@AkihiroSuda, @stevvooe I just rebased PR, because I need the latest changes in order to re-vendor the docker/docker repo

aluzzardi commented:
@stevvooe Are you okay with merging this?

if so, @vasil-yordanov, could you rebase the PR once more?

stevvooe commented:
@aluzzardi Mechanically, this PR is fine and meets the use case that @vasil-yordanov is trying to solve. It will basically set the hostname to the value of this field for all replicas of the service.

The question is how will this behave in a more interesting scheme, where hostnames are the result of some calculation of service state (ie slot id, alias pool, etc.). The other aspect is the level of network integration that is expected. Typically, this hostname value only has an impact on how the container process see the hostname, but there may be expectations that this also sets the network hostname. While this is pedantically the correct behavior, it is not the expected behavior for newer users, who are unlikely to have the sophistication to understand the subtle differences (and I still don't fully understand their implications).

My worry is that people will set the hostname and get exactly what this feature provides but still not be happy. We need to consider the implications.

I would love to see a complete proposal from the networking team. @mrjana @mavenugo

aluzzardi commented:
@vasil-yordanov Could you please rebase one more time so we can get this merged?

Thanks for keeping up with this, I know it's been a while!

vasil-yordanov commented:
@aluzzardi I have rebased the code. Thanks for merging this.

GordonTheTurtle commented:
Looks like we would not be able to merge this PR because of merge conflicts. Please rebase, fix conflicts, and force push to your branch.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "master" git@github.com:vasil-yordanov/swarmkit.git somewhere
$ cd somewhere
$ git rebase -i HEAD~95
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

stevvooe commented:
@vasil-yordanov I think something weird happened with the rebase. May want to try again.

Also, please check out #1650. I think we can add support for hostname and meet the entire feature request.

niau commented:
@stevvooe , @GordonTheTurtle  sorry about the mess. Vassil was busy with other stuff and I have tried to help on his behalf. Will try to correct it as per your proposal.  
Just as a question won't it be easier to try to simply revert my commits?

niau commented:
Anyway for now I have amended my commits so that they are signed off. On monday will try to drop them and rebase one more time. If sucesfull will try to figure out how can we benefit from the templates in #1650.

vasil-yordanov commented:
@stevvooe I'm sorry but currently I'm very busy and cannot contribute much. I'm sure @niau will manage to rebase. I'm in constant contact with him.

stevvooe commented:
Carried in #1688.
